File: src/main/java/com/cse3063f25grp1/writer/HeuristicQueryWriter.java
package com.cse3063f25grp1.writer;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import com.cse3063f25grp1.model.Intent;

public class HeuristicQueryWriter implements QueryWriter {

    private static final Locale TR_LOCALE = new Locale("tr", "TR");
    private static final String SPLIT_REGEX = "\\s+";
    private static final String CLEAN_REGEX = "[^\\p{L}0-9 ]";
    private final Set<String> stopwords;
    private final Map<Intent, List<String>> intentBoosters;

    public HeuristicQueryWriter(Set<String> stopwords, Map<Intent, List<String>> intentBoosters) {
        if (stopwords == null) {
            this.stopwords = Collections.emptySet();
        } else {
            this.stopwords = stopwords.stream()
                    .filter(s -> s != null && !s.isBlank())
                    .map(s -> s.toLowerCase(TR_LOCALE))
                    .collect(Collectors.toUnmodifiableSet());
        }
        if (intentBoosters == null) {
            this.intentBoosters = Collections.emptyMap();
        } else {
            Map<Intent, List<String>> normalized = new LinkedHashMap<>();
            for (Map.Entry<Intent, List<String>> entry : intentBoosters.entrySet()) {
                List<String> normalizedTokens = entry.getValue().stream()
                        .filter(s -> s != null && !s.isBlank())
                        .map(s -> s.toLowerCase(TR_LOCALE))
                        .collect(Collectors.toUnmodifiableList());
                normalized.put(entry.getKey(), normalizedTokens);
            }
            this.intentBoosters = Collections.unmodifiableMap(normalized);
        }
    }

    @Override
    public List<String> write(String question, Intent intent) {
        if (question == null || question.isBlank()) {
            return Collections.emptyList();
        }

        String cleaned = question.toLowerCase(TR_LOCALE).replaceAll(CLEAN_REGEX, " ");
        List<String> parts = Arrays.asList(cleaned.split(SPLIT_REGEX));

        LinkedHashSet<String> orderedTerms = new LinkedHashSet<>();
        for (String p : parts) {
            String token = p.trim();
            if (token.isEmpty()) {
                continue;
            }
            if (stopwords.contains(token)) {
                continue;
            }
            orderedTerms.add(token);
        }

        Intent resolvedIntent = intent == null ? Intent.Unknown : intent;
        List<String> boosters = intentBoosters.get(resolvedIntent);
        if (boosters != null) {
            for (String booster : boosters) {
                orderedTerms.add(booster);
            }
        }

        return new ArrayList<>(orderedTerms);
    }
}

File: iteration1_py/src/writer/heuristic_query_writer.py
import re
from typing import Dict, List, Set, Optional

from src.model.intent import Intent
from src.writer.query_writer_abstract import QueryWriter


class HeuristicQueryWriter(QueryWriter):


    # TR locale pattern is approximated because Python locale lowercasing differs
    CLEAN_REGEX = r"[^A-Za-zÇĞİÖŞÜçğıöşü0-9 ]"
    SPLIT_REGEX = r"\s+"

    def __init__(
        self,
        stopwords: Optional[Set[str]],
        intent_boosters: Optional[Dict[Intent, List[str]]]
    ) -> None:

        # Normalize stopwords
        if stopwords is None:
            self.__stopwords: Set[str] = set()
        else:
            self.__stopwords = {
                s.lower() for s in stopwords
                if s is not None and s.strip() != ""
            }

        # Normalize boosters
        if intent_boosters is None:
            self.__intent_boosters: Dict[Intent, List[str]] = {}
        else:
            normalized: Dict[Intent, List[str]] = {}

            for intent, tokens in intent_boosters.items():
                clean_tokens: List[str] = [
                    t.lower()
                    for t in tokens
                    if t is not None and t.strip() != ""
                ]
                normalized[intent] = clean_tokens

            self.__intent_boosters = normalized

    
    def write(self, question: str, intent: Intent) -> List[str]:
        

        if question is None or question.strip() == "":
            return []

        # Lowercase & clean unwanted characters
        lower: str = question.lower()
        cleaned: str = re.sub(self.CLEAN_REGEX, " ", lower)

        # Tokenize
        parts: List[str] = re.split(self.SPLIT_REGEX, cleaned)

        # Ordered unique tokens (LinkedHashSet equivalent)
        ordered_terms: dict[str, None] = {}

        for p in parts:
            token: str = p.strip()
            if not token:
                continue
            if token in self.__stopwords:
                continue
            ordered_terms[token] = None

        # Add boosters according to detected intent
        resolved_intent: Intent = intent if intent is not None else Intent.UNKNOWN
        boosters: List[str] = self.__intent_boosters.get(resolved_intent, [])

        for booster in boosters:
            ordered_terms[booster] = None

        # Convert back to ordered list
        return list(ordered_terms.keys())
